api/src/main/java/org/apache/iceberg/SortOrderComparators.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.lang.reflect.Array;
+import java.util.Comparator;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+public class SortOrderComparators {
+  private SortOrderComparators() {}
+
+  /** Compare structs with the specified sort order projection */
+  public static Comparator<StructLike> forSchema(Schema schema, SortOrder sortOrder) {
+    Preconditions.checkArgument(sortOrder.isSorted(), "Invalid sort order: unsorted");
+    SortOrder.checkCompatibility(sortOrder, schema);
+    return new SortOrderComparator(schema, sortOrder);
+  }
+
+  /** Util method to chain sort direction and null order to the original comparator. */
+  private static Comparator<Object> sortFieldComparator(
+      Comparator<Object> original, SortField sortField) {
+    Comparator<Object> comparator = original;
+    if (sortField == null) {
+      return Comparators.nullsFirst().thenComparing(comparator);
+    }
+
+    if (sortField.direction() == SortDirection.DESC) {
+      comparator = comparator.reversed();
+    }
+
+    if (sortField.nullOrder() == NullOrder.NULLS_FIRST) {
+      comparator = Comparators.nullsFirst().thenComparing(comparator);
+    } else if (sortField.nullOrder() == NullOrder.NULLS_LAST) {
+      comparator = Comparators.nullsLast().thenComparing(comparator);
+    }
+
+    return comparator;
+  }
+
+  private static class SortOrderComparator implements Comparator<StructLike> {
+    private final SortKey leftKey;
```

Comment:
included `SortOrderComparator` impl in this PR to demonstrate the usage or `SortKey`